### PR TITLE
Add a rich console exporter

### PR DESCRIPTION
**Pull request by tonybaloney**
# Description

This is an alternative to the Console Exporter, that uses the Rich Console library to print colorised trees from a batch span processor. Span attributes are shown as nodes in the tree and spans are children of their related parent span/trace.

This is easier to read than the console exporter and see the relationships between traces.

Useful for debugging.

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Test A

# Does This PR Require a Core Repo Change?

- [ ] Yes. - Link to PR: 
- [ ] No.

# Checklist:

See [contributing.md](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/main/CONTRIBUTING.md) for styleguide, changelog guidelines, and more.

- [ ] Followed the style guidelines of this project
- [ ] Changelogs have been updated
- [ ] Unit tests have been added
- [ ] Documentation has been updated

Example output:

```
Trace 6d3953ac92706f9dcbdc7129a4c72736                                                                                              
└── [01:40:55.139608] /locations, span 503921bf52866810                                                                             
    ├── Kind : SERVER                                                                                                               
    ├── Attributes :                                                                                                                
    │   ├── http.scheme : http                                                                                                      
    │   ├── http.host : 127.0.0.1:8000                                                                                              
    │   ├── net.host.port : 8000                                                                                                    
    │   ├── http.flavor : 1.1                                                                                                       
    │   ├── http.target : /locations                                                                                                
    │   ├── http.url : http://127.0.0.1:8000/locations                                                                              
    │   ├── http.method : GET                                                                                                       
    │   ├── http.server_name : 127.0.0.1:8000                                                                                       
    │   ├── http.user_agent : Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/605.1.15 (KHTML, like Gecko)              
    │   │   Version/14.1.2 Safari/605.1.15                                                                                          
    │   ├── net.peer.ip : 127.0.0.1                                                                                                 
    │   ├── net.peer.port : 64673                                                                                                   
    │   ├── http.route : /locations                                                                                                 
    │   └── http.status_code : 200                                                                                                  
    ├── [01:40:55.142977] SELECT "id","name","private" FROM "locations", span c12862a73454ecc1                                      
    │   ├── Kind : CLIENT                                                                                                           
    │   └── Attributes :                                                                                                            
    │       ├── db.system : sqlite                                                                                                  
    │       ├── db.name : sample.db                                                                                                 
    │       ├── db.statement :                                                                                                      
    │       └── SELECT "id","name","private" FROM "locations"                                                                       
    ├── [01:40:55.145724] /locations http send, span 08dff3718cf0a4a3                                                               
    │   ├── Kind : INTERNAL                                                                                                         
    │   └── Attributes :                                                                                                            
    │       ├── http.status_code : 200                                                                                              
    │       └── type : http.response.start                                                                                          
    └── [01:40:55.147593] /locations http send, span d3088e87ff397dc4                                                               
        ├── Kind : INTERNAL                                                                                                         
        └── Attributes :                                                                                                            
            └── type : http.response.body                                 
```

Colorised:

screenshot 2021-09-18 at 13 16 51

### CODE REVIEW

**Comment by owais** [comment]
Thanks. This looks very nice and useful but is this something that the OpenTelemetry project should host and maintain? This could easily be hosted independently and published as a 3rd party package to pypi.org. Curious what other @open-telemetry/python-approvers @open-telemetry/python-maintainers think about it.

**Comment by owais** [comment]
If we were to host it in this repo, I think it'd be better to add the feature to the existing console exporter and make it configurable to print either format.

**Comment by tonybaloney** [comment]
> If we were to host it in this repo, I think it'd be better to add the feature to the existing console exporter and make it configurable to print either format.

Happy to host + publish this one myself. Can I use the same package name (`opentelemetry-exporter-richconsole`) ?

If it were added to the console exporter, you'd have to add Rich as a dependency to the console exporter, which would be overkill IMO

**Comment by lonewolf3739** [comment]
> If it were added to the console exporter, you'd have to add Rich as a dependency to the console exporter, which would be overkill IMO

Making it optional dependency with `extras_require` and use rich format if it exists?

**Comment by owais** [comment]
> Can I use the same package name (opentelemetry-exporter-richconsole) ?

I think so. Unfortunately pypi doesn't have namespaces like NPM and I don't think we can enforce `opentelemetry-` prefix not being used by others so eventually packages will pop up with that prefix. I think it's fine.

**Comment by aabmass** [comment]
> This looks very nice and useful but is this something that the OpenTelemetry project should host and maintain?

@owais how does this differ from other exporters/instrumentations in this repo? Trying to understand arguments for and against this being merged here.

**Comment by owais** [comment]
@aabmass I'm not sure either and that is why I'm bringing it up. I don't think there is anything inherently wrong with it being in contrib but at the same time every new package adds additional burden on maintainers and systems (CI) both. It is usually worth it for packages that improve the quality of telemetry this project can provide (instrumentations) but not sure if every dev/debugging utility needs to be hosted in this repo. May be we can encourage the broader OpenTelemetry (Python) ecosystem to grow beyond core and contrib repos. I don't feel strongly either way but felt it was an interesting this to talk about.

**Comment by aabmass** [comment]
@owais I think this one adds a lot of value for developers (both instrumentation authors and application developers). I'm in favor of adding this, but I agree it's worth a general discussion. I've added this PR to the agenda for next week's SIG.

**Review by owais — APPROVED** [review]

**Comment by owais** [comment]
@tonybaloney please fix lint issues, add a changelog entry and we can merge this.

File: exporter/opentelemetry-exporter-richconsole/src/opentelemetry/exporter/richconsole/__init__.py
Comment (aabmass) [inline exporter/opentelemetry-exporter-richconsole/src/opentelemetry/exporter/richconsole/__init__.py:144]: no need to copy the spans list

```suggestion
        for span in spans:
            if not span.parent:
                ...
```

File: exporter/opentelemetry-exporter-richconsole/src/opentelemetry/exporter/richconsole/__init__.py
Comment (aabmass) [inline exporter/opentelemetry-exporter-richconsole/src/opentelemetry/exporter/richconsole/__init__.py:155]: same comment as above
```suggestion
        for span in spans:
            if span.parent and span.parent.span_id not in parents:
```

**Review by aabmass — CHANGES_REQUESTED** [review]
A few comments, otherwise LGTM

File: exporter/opentelemetry-exporter-richconsole/README.rst
Comment (tonybaloney) [reply exporter/opentelemetry-exporter-richconsole/README.rst:17]: It could do, I haven't explored metrics exporters yet. Will leave it open

File: exporter/opentelemetry-exporter-richconsole/src/opentelemetry/exporter/richconsole/__init__.py
Comment (lonewolf3739) [inline exporter/opentelemetry-exporter-richconsole/src/opentelemetry/exporter/richconsole/__init__.py:78]: Why is this necessary? Does it not work with `SimpleSpanProcessor`?

File: exporter/opentelemetry-exporter-richconsole/src/opentelemetry/exporter/richconsole/__init__.py
Comment (lonewolf3739) [inline exporter/opentelemetry-exporter-richconsole/src/opentelemetry/exporter/richconsole/__init__.py:149]: Can we remove this?

**Review by lonewolf3739 — APPROVED** [review]
LGTM. Added couple of comments.

**Comment by tonybaloney** [comment]
Whats the correct way to run the linter to get the rest of this to pass?

I tried running black but it reformats everything, tried `scripts/eachdist.py lint` and that reformatted everything as well

**Comment by owais** [comment]
@tonybaloney make sure your installed black and isort versions are the same as pinned in dev-requirements.txt, then run `./scripts/eachdist.py format`

File: exporter/opentelemetry-exporter-richconsole/setup.cfg
Comment (codeboten) [inline exporter/opentelemetry-exporter-richconsole/setup.cfg:33]: this should probably include python 3.9 as well

File: exporter/opentelemetry-exporter-richconsole/src/opentelemetry/exporter/richconsole/__init__.py
Comment (codeboten) [inline exporter/opentelemetry-exporter-richconsole/src/opentelemetry/exporter/richconsole/__init__.py:32]: Just confirming that using it with a different processor will still work. If that's the case, maybe `must` can be changed to `should`

**Review by codeboten — COMMENTED** [review]
Just a couple of comments, this is a good debugging tool!

**Comment by tonybaloney** [comment]
@codeboten done and done!

**Review by ocelotl — APPROVED** [review]

**Review by codeboten — APPROVED** [review]
Thanks for addressing my comments!

**Comment by codeboten** [comment]
@aabmass please approve if your comments have been addressed

**Review by aabmass — APPROVED** [review]
🚢

**Pull request merged**
